Userland/Libraries/LibJS/Runtime/Temporal/ISO8601.cpp

```diff
@@ -0,0 +1,505 @@
+/*
+ * Copyright (c) 2021, Linus Groh <linusg@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/CharacterTypes.h>
+#include <LibJS/Runtime/Temporal/ISO8601.h>
+
+namespace JS::Temporal {
+
+namespace Detail {
+
+#define TRY_PARSE(scope)                            \
+    ({                                              \
+        save_state();                               \
+        auto start = m_state.lexer.tell();          \
+        scope;                                      \
+        auto end = m_state.lexer.tell();            \
+        discard_saved_state();                      \
+        m_input.substring_view(start, end - start); \
+    })
+
+#define PARSE_REQUIRED(expression) \
+    ({                             \
+        if (!(expression)) {       \
+            restore_saved_state(); \
+            return false;          \
+        }                          \
+        true;                      \
+    })
+
+#define PARSE_OPTIONAL(expression)           \
+    ({                                       \
+        save_state();                        \
+        (!(expression))                      \
+            ? (restore_saved_state(), false) \
+            : (discard_saved_state(), true); \
+    })
+
+// https://tc39.es/proposal-temporal/#prod-DecimalDigit
+bool ISO8601Parser::parse_decimal_digit()
+{
+    // DecimalDigit : one of
+    //     0 1 2 3 4 5 6 7 8 9
+    TRY_PARSE({
+        PARSE_REQUIRED(m_state.lexer.next_is(is_ascii_digit));
+        m_state.lexer.consume();
+    });
+    return true;
```

Comment:
(continuing from above)
```suggestion
    if (m_state.lexer.next_is(is_ascii_digit)) {
        m_state.lexer.consume();
        return true;
    }
    return false;
```